### PR TITLE
use proper ErrorCode in table function Value

### LINKED ISSUE #43331: Logical error: 'Table function 'values' must have arguments'.
Sending an incorrect query (missing the table types):

```
$ clickhouse client --query "SELECT * FROM VALUES    (1)
    (2)
    (3)"
Received exception from server (version 22.10.2):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Table function 'values' must have arguments. (LOGICAL_ERROR)
(query: SELECT * FROM VALUES    (1)
    (2)
    (3))
```

```
2022.11.17 16:04:23.000709 [ 247143 ] {22b0298f-51eb-49c9-8100-d2e1cd9c4bfa} <Error> executeQuery: Code: 49. DB::Exception: Table function 'values' must have arguments. (LOGICAL_ERROR) (version 22.10.2.11 (official build)) (from [::ffff:127.0.0.1]:49124) (in query: SELECT * FROM VALUES (1) (2) (3)), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xce3f77a in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
1. DB::TableFunctionValues::parseArguments(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>) @ 0x111ad823 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
2. DB::TableFunctionFactory::get(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>) const @ 0x11494e47 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
3. DB::Context::executeTableFunction(std::__1::shared_ptr<DB::IAST> const&) @ 0x11f01bc0 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
4. DB::JoinedTables::getLeftTableStorage() @ 0x126aae90 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
5. DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context> const&, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::PreparedSets>) @ 0x125d5a32 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
6. DB::InterpreterSelectWithUnionQuery::buildCurrentChildInterpreter(std::__1::shared_ptr<DB::IAST> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x126715c2 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
7. DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x1266f340 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
8. DB::InterpreterFactory::get(std::__1::shared_ptr<DB::IAST>&, std::__1::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x125969fe in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
9. ? @ 0x12979320 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
10. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x12976c0d in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
11. DB::TCPHandler::runImpl() @ 0x1353dd04 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
12. DB::TCPHandler::run() @ 0x13551cf9 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
13. Poco::Net::TCPServerConnection::start() @ 0x1636274f in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
14. Poco::Net::TCPServerDispatcher::run() @ 0x16364adb in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
15. Poco::PooledThread::run() @ 0x16520372 in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
16. Poco::ThreadImpl::runnableEntry(void*) @ 0x1651dafd in /mnt/ch/official_binaries/clickhouse-common-static-22.10.2.11/usr/bin/clickhouse
17. ? @ 0x7f1b3b5f38fd in ?
18. ? @ 0x7f1b3b675a60 in ?
```

This needs a better error.

### PR DESCRIPTION
**Pull request by ucasfl**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Closes #43331.

### CODE REVIEW

**Review by kssenii — APPROVED** [review]

File: src/TableFunctions/TableFunctionValues.cpp [hunk 1]
```diff
@@ -83,7 +82,7 @@ void TableFunctionValues::parseArguments(const ASTPtr & ast_function, ContextPtr
     ASTs & args_func = ast_function->children;
 
     if (args_func.size() != 1)
-        throw Exception("Table function '" + getName() + "' must have arguments", ErrorCodes::LOGICAL_ERROR);
+        throw Exception("Table function '" + getName() + "' must have arguments", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
```
Comment (tavplubix) [inline src/TableFunctions/TableFunctionValues.cpp:85]: First of all, why not to replace it with formatted message like this?
```suggestion
        throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Table function '{}' must have arguments", getName());
```

File: src/TableFunctions/TableFunctionValues.cpp [hunk 2]
```diff
@@ -83,7 +82,7 @@ void TableFunctionValues::parseArguments(const ASTPtr & ast_function, ContextPtr
     ASTs & args_func = ast_function->children;
 
     if (args_func.size() != 1)
-        throw Exception("Table function '" + getName() + "' must have arguments", ErrorCodes::LOGICAL_ERROR);
```
Comment (tavplubix) [inline src/TableFunctions/TableFunctionValues.cpp:86]: Second, it's not the only place where exception like this is thrown. Just grep `must have arguments", ErrorCodes::LOGICAL_ERROR` and you will find at least 2 more places. Should we fix other places as well?

**Review by tavplubix — CHANGES_REQUESTED** [review]
And third: what about adding some tests?

**Comment by tavplubix** [comment]
And finally: why do you think that it's not a logical error?

Is it possible to have a function without arguments list? Of course it's possible to have a function with empty arguments list (like `func()`), but not without arguments list (`func` is an identifier, not a function). 

Seems like behavior of `ParserFunction` was changed recently (we can ask @evillique), so now the parser can produce invalid AST. I guess we have many places in code which assume that `ASTFunction` always has one child and it's `ASTExpressionList`. We have to either find all such places and fix them or fix `ParserFunction` and keep `LOGICAL_ERROR` here.

**Comment by evillique** [comment]
There was a little [bug](https://github.com/ClickHouse/ClickHouse/pull/43350/files#diff-c9b1b8a9340030064ff0f78deb256c3d34608af5b2d4a5bd0a59eee7260733e8R923) in the ExpressionParser that allowed it to parse an unlimited amount of round brackets into one function, but the same error [occurs](https://fiddle.clickhouse.com/7384d943-3d73-4cfb-a5a5-dcf1e6942a56) on the 22.9 version (before the new parser) where there are two pairs of brackets.

The actual reason it happens is not that `args_func.size()` is 0, but because it actually equals 2. When we parse a function through ParserFunction we can specify if we want two pairs of brackets to be parsed as `parameters` and `arguments` in ASTFunction respectively. In this case, we will put both of them in the children of ASTFunction and get this error.

It is used in aggregate functions, but here I don't see any table functions that use it, so I tried to disable it in the previously mentioned PR as well.

**Comment by ucasfl** [comment]
> And finally: why do you think that it's not a logical error?
> 
> Is it possible to have a function without arguments list? Of course it's possible to have a function with empty arguments list (like `func()`), but not without arguments list (`func` is an identifier, not a function).
> 
> Seems like behavior of `ParserFunction` was changed recently (we can ask @evillique), so now the parser can produce invalid AST. I guess we have many places in code which assume that `ASTFunction` always has one child and it's `ASTExpressionList`. We have to either find all such places and fix them or fix `ParserFunction` and keep `LOGICAL_ERROR` here.

It's ok to keep it and fix `ParserFunction`, and many other table functions also should fix it, because currently lot of them use `NUMBER_OF_ARGUMENTS_DOESNT_MATCH` as the ErrorCode.

**Comment by tavplubix** [comment]
Closed in favor of https://github.com/ClickHouse/ClickHouse/pull/43350

**Pull request closed without merging**